sstables/sstable_set.cc

```diff
@@ -768,6 +768,393 @@ sstable_set_impl::create_single_key_sstable_reader(
     return make_combined_reader(schema, std::move(permit), std::move(readers), fwd, fwd_mr);
 }
 
+// Merges output from readers of TWCS sstables opened for a single partition query into a non-decreasing stream
+// of mutation fragments.
+//
+// Uses `min_position_reader_queue` to open the readers lazily as the read progresses through the partition.
+// A reader is opened only if we find that it may contain fragments for the currently inspected positions.
+//
+// Readers are closed as soon as we find that they were exhausted for the given partition query.
+//
+// Implements the `FragmentProducer` concept. However, `next_partition` and `fast_forward_to(partition_range)`
+// are not implemented and throw an error; the reader is only used for single partition queries.
+//
+// Assumes that the readers returned from `min_position_reader_queue` return fragments from the same partition.
+// Otherwise the behavior is unspecified.
+class time_series_reader_merger {
+    const schema_ptr _schema;
+    const reader_permit _permit;
+
+    // The key of the partition being queried.
+    const dht::decorated_key _key;
+
+    // Compares positions using *_schema.
+    const position_in_partition::less_compare _less;
+    const position_in_partition::equal_compare _eq;
+
+    // A queue of readers used to lazily open new readers as we progress through the partition.
+    // Before a batch with position `p` is returned, we ensure that all readers containing positions
+    // <= `p` are popped from the queue.
+    min_position_reader_queue _reader_queue;
+
+    // Owning container for the readers popped from _reader_queue.
+    // If we are sure that a reader is exhausted (all rows from the queried partition have been returned),
+    // we destroy and remove it from the container.
+    std::list<backed_reader> _all_readers;
+    using reader_iterator = std::list<backed_reader>::iterator;
+
+    // A reader together with the last fragment it returned.
+    // The iterator points to _all_readers.
+    // fragment.position() is <= than position of the next fragment returned by reader, if any.
+    struct peeked_reader {
+        reader_iterator reader;
+        mutation_fragment fragment;
+    };
+
+    // A min-heap of peeked readers, sorted by the position in partition of the last returned fragment.
+    merger_vector<peeked_reader> _peeked_readers;
```

Comment:
What is a `merger_vector<>`?

Reply:
Forgot to include the appropriate header in this commit, fixd